### ISSUE TITLE
[ibm-telco] Issues observed while running ztp playbooks

### ISSUE THREAD

**Issue by maheshd2** [redhat-ztp/ztp-cluster-deploy#33]
Hi yolanda,

Also as discussed, here are the issues we faced while running ztp playbooks:

## Issues with HIGH priority

1. httpd service was found stopped after "prepare-environment" step. We manually started it.

2. In "racadm.sh", Resolving the host IP address for http service to be enhanced to pick right IP. The command was returning non-routable IP address which was not reachable by worker node.

3. Root partition space issue. The installation will always use root partition to place virtual machine images. Even if we make changes in libvirt storage pool to point to a different path, Playbooks will delete the storage pool. So, it is necessory to have enough space under root partition. Or is there any way to place VM images outside root partition..?

4. Virtual Media boot fails intermittently on dell servers. It is observed that once the CoreOS is booted, it prevents subsequent booting of new OS via Virtual Media. So is it something that needs to be taken care on BIOS settings.?

## Issues with MEDIUM priority

5. Many dependent packages, modules missing and unset/undefined variables observed. I have all the missing ones documented in the ztp_install_guide.md. Need detailed guide on HOST preparation in terms of packages/bios settings etc. Most of them can be added to the playbooks.

6. k8s.gcr.io/pause:3.2 was not showing up after "prepare-environment" step. We ignored and went ahead.

## Issues with LOW priority

7. Finally worker node addition is failing due to "API VIP connectivity failure". Able to ping the VIP address rom CONTROL Plain host as well as from worker node.
   [SOLVED - There was an issue with name resolution for the api_vip from worker node]

8. Some times the UI goes down. Not sure if this is because of space issue.

9. Once installation failed due to "pull rate limit" while running "expose-endpoints" step. This is intermittent and something to deal with redhat subscription account.

10. Sometimes intermittently installation failed with missing attribute. Not sure what was the cause. Has to cleanup environment and re-deploy to solve this.

11. Sometime also found issue with racadm.sh binary, when I tried to modify and run.

12. How to add more worker nodes not specified. We see syntax for just adding one worker node.

13. I found these IP address defined in the inventory hosts, was not used/assigned. Instead they got the IP from dhcp server. Will DHCP will take precedence.?

### Issue 1. 
httpd service was found stopped after "prepare-environment" step. I manually started it.

```
[root@ztphost ai-deploy-cluster-remoteworker]# sudo systemctl status httpd
● httpd.service - The Apache HTTP Server
   Loaded: loaded (/usr/lib/systemd/system/httpd.service; enabled; vendor preset: disabled)
   Active: inactive (dead)
     Docs: man:httpd.service(8)
``` 

### Issue 2. 
Resolving the host IP address for http service. The command was returning non-routable IP address which was not reachable by worker node. Made below change to address this issue. 

```
Edit "roles/add-remote-worker/templates/racadm.sh"

Replace:
$(hostname -I | cut -f2 -d" " | xargs)
With:
$(hostname -i | xargs)


And verify the IP is resolved to right one.
[rhacm@ztphost html]$ cat /etc/hosts | grep 9.42.98.150
9.42.98.150  ztphost.rtp.raleigh.ibm.com
[rhacm@ztphost html]$ hostname -i
9.42.98.150
```

```
fatal: [localhost]: FAILED! => {
    "changed": true,
    "cmd": "/root/racadm.sh root 3KWAAMK9W3PN 9.42.98.136 coreos.iso",
    "delta": "0:00:17.750476",
    "end": "2021-01-20 02:29:59.984161",
    "invocation": {
        "module_args": {
            "_raw_params": "/root/racadm.sh root 3KWAAMK9W3PN 9.42.98.136 coreos.iso",
            "_uses_shell": true,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "msg": "non-zero return code",
    "rc": 1,
    "start": "2021-01-20 02:29:42.233685",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "Security Alert: Certificate is invalid - self signed certificate\nContinuing execution. Use -S option for racadm to stop execution on certificate-related errors.\n\r                                                                             \r\r                                                                             \rRemote File Share is Enabled\r\nUserName \r\nPassword \r\nShareName http://9.46.66.245/rhel-8.2-x86_64-dvd.iso\r\n\nSecurity Alert: Certificate is invalid - self signed certificate\nContinuing execution. Use -S option for racadm to stop execution on certificate-related errors.\n\r                                                                             \r\r                                                                             \rDisable Remote File Started. Please check status using -s\noption to know Remote File Share is ENABLED or DISABLED.\r\n\nSecurity Alert: Certificate is invalid - self signed certificate\nContinuing execution. Use -S option for racadm to stop execution on certificate-related errors.\n\r                                                                             \r\r                                                                             \rERROR: Unable to perform requested operation.",
    "stdout_lines": [
        "Security Alert: Certificate is invalid - self signed certificate",
        "Continuing execution. Use -S option for racadm to stop execution on certificate-related errors.",
        "",
        "                                                                             ",
        "",
        "                                                                             ",
        "Remote File Share is Enabled",
        "UserName ",
        "Password ",
        "ShareName http://9.46.66.245/rhel-8.2-x86_64-dvd.iso",
        "",
        "Security Alert: Certificate is invalid - self signed certificate",
        "Continuing execution. Use -S option for racadm to stop execution on certificate-related errors.",
        "",
        "                                                                             ",
        "",
        "                                                                             ",
        "Disable Remote File Started. Please check status using -s",
        "option to know Remote File Share is ENABLED or DISABLED.",
        "",
        "Security Alert: Certificate is invalid - self signed certificate",
        "Continuing execution. Use -S option for racadm to stop execution on certificate-related errors.",
        "",
        "                                                                             ",
        "",
        "                                                                             ",
        "ERROR: Unable to perform requested operation."
    ]
}
```

### Issue 3. 
Root partition space issue. The installation will always use root partition to place virtual machine images. Even if we make changes in libvirt storage pool to point to a different path, Playbooks will delete the custom storage pool. So, it is necessory to have enough space under root partition.

```
    "stderr": "Traceback (most recent call last):\n  File \"/bin/aicli\", line 11, in <module>\n    load_entry_point('aicli==99.0', 'console_scripts', 'aicli')()\n  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/cli.py\", line 423, in cli\n    args.func(args)\n  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/cli.py\", line 137, in download_iso\n    ai.download_iso(args.cluster, args.path)\n  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/__init__.py\", line 156, in download_iso\n    copyfileobj(response, f)\n  File \"/usr/lib64/python3.6/shutil.py\", line 82, in copyfileobj\n    fdst.write(buf)\nOSError: [Errno 28] No space left on device",
    "stderr_lines": [
        "Traceback (most recent call last):",
        "  File \"/bin/aicli\", line 11, in <module>",
        "    load_entry_point('aicli==99.0', 'console_scripts', 'aicli')()",
        "  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/cli.py\", line 423, in cli",
        "    args.func(args)",
        "  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/cli.py\", line 137, in download_iso",
        "    ai.download_iso(args.cluster, args.path)",
        "  File \"/usr/local/lib/python3.6/site-packages/aicli-99.0-py3.6.egg/ailib/__init__.py\", line 156, in download_iso",
        "    copyfileobj(response, f)",
        "  File \"/usr/lib64/python3.6/shutil.py\", line 82, in copyfileobj",
        "    fdst.write(buf)",
        "OSError: [Errno 28] No space left on device"
    ],
    "stdout": "\u001b[0;32mDownloading Iso for Cluster ztp-day2 in /opt/cached_disconnected_images\u001b[0;0m",
    "stdout_lines": [
        "\u001b[0;32mDownloading Iso for Cluster ztp-day2 in /opt/cached_disconnected_images\u001b[0;0m"
    ]
```

### Issue 4. 
Virtual Media boot fails intermittently on dell servers. It is observed that once the CoreOS is booted, it prevents subsequent booting o new OS via Virtual Media. So is it something that needs to be taken care on BIOS settings.?

### Issue 5. 
Installation failed many times due to missing of many dependent packages, modules and unset/undefined variables. I have all the missing ones documented in the issue https://github.com/redhat-ztp/ztp-cluster-deploy/issues/32 . Also created soft links for some of the binaries.

### Issue 6. 
k8s.gcr.io/pause:3.2 was not showing up after "prepare-environment" step. Ignored and went ahead.

```
[root@ztphost ai-deploy-cluster-remoteworker]# sudo podman ps
CONTAINER ID  IMAGE                                          COMMAND               CREATED         STATUS             PORTS                   NAMES
15344f7b0ec0  quay.io/ocpmetal/assisted-service:latest       /assisted-service     23 seconds ago  Up 22 seconds ago  0.0.0.0:5432->5432/tcp  installer
0f1d33ef01ad  quay.io/ocpmetal/ocp-metal-ui:latest           /opt/bitnami/scri...  24 seconds ago  Up 24 seconds ago  0.0.0.0:5432->5432/tcp  ui
126068ba9f32  quay.io/ocpmetal/postgresql-12-centos7:latest  run-postgresql        26 seconds ago  Up 26 seconds ago  0.0.0.0:5432->5432/tcp  db
```

### Issue 7. 
Finally worker node addition is failing due to "API VIP connectivity failure"
[SOLVED - There was an issue in name resolution for api vip address from the worker node]

### Issue 8. Some times the UI goes down. (This may be due to space/disk issue, which I observed only once)

```
[root@ztphost ai-deploy-cluster-remoteworker]# netstat -anlp | grep 8080
tcp        3      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      108884/conmon
tcp      202      0 9.42.98.150:8080        9.42.98.150:43858       CLOSE_WAIT  -
tcp      202      0 9.42.98.150:8080        9.42.98.150:43710       CLOSE_WAIT  -
tcp      202      0 9.42.98.150:8080        9.42.98.150:45150       CLOSE_WAIT  -
```

### Issue 9. 
Once installation failed due to "pull rate limit" while running "expose-endpoints" step (This was observed only once, may be something related to redhat subscription user)
```
    "msg": "Can't pull image docker.io/haproxy:1.7",
    "stderr": "Trying to pull docker.io/haproxy:1.7...\ntime=\"2021-01-19T13:29:51-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"\ntime=\"2021-01-19T13:29:53-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"\ntime=\"2021-01-19T13:29:57-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"\ntime=\"2021-01-19T13:30:05-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"\n  toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit\nError: error pulling image \"docker.io/haproxy:1.7\": unable to pull docker.io/haproxy:1.7: unable to pull image: Error initializing source docker://haproxy:1.7: Error reading manifest 1.7 in docker.io/library/haproxy: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit\n",
    "stderr_lines": [
        "Trying to pull docker.io/haproxy:1.7...",
        "time=\"2021-01-19T13:29:51-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"",
        "time=\"2021-01-19T13:29:53-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"",
        "time=\"2021-01-19T13:29:57-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"",
        "time=\"2021-01-19T13:30:05-05:00\" level=error msg=\"HEADER map[Cache-Control:[no-cache] Content-Type:[application/json]]\"",
        "  toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit",
        "Error: error pulling image \"docker.io/haproxy:1.7\": unable to pull docker.io/haproxy:1.7: unable to pull image: Error initializing source docker://haproxy:1.7: Error reading manifest 1.7 in docker.io/library/haproxy: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit"
    ],
    "stdout": "",
    "stdout_lines": []
```

### Issue 10. 
Sometimes intermittently installation failed with missing attribute. (This was observed only once, not sure what was the reason)
```
TASK [add-remote-worker : Wait until hosts are ready in Assisted Installer] ******************************************************************************
task path: /home/rhacm/ztp/ztp-cluster-deploy-master/ai-deploy-cluster-remoteworker/roles/add-remote-worker/tasks/main.yml:103
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716 `" && echo ansible-tmp-1611136010.454818-56004-128519372912716="` echo /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/commands/command.py
<localhost> PUT /root/.ansible/tmp/ansible-local-55724u73ifus4/tmp3ul9xi_9 TO /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716/AnsiballZ_command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716/ /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c 'AI_URL=http://9.42.98.150:8080 /usr/libexec/platform-python /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1611136010.454818-56004-128519372912716/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "msg": "The conditional check 'hosts_ready.stdout|int == 1' failed. The error was: error while evaluating conditional (hosts_ready.stdout|int == 1): 'dict object' has no attribute 'stdout'"
}
```

### Issue 11. 
Sometime also found issue with racadm.sh binary, when I tried to modify and run. (This was observed only once, not sure what was the reason)
```
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1611133059.4247096-45425-255428948790796/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": true,
    "cmd": "/root/racadm.sh root 3KWAAMK9W3PN 9.42.98.136 coreos.iso",
    "delta": "0:00:00.005584",
    "end": "2021-01-20 03:57:39.716135",
    "invocation": {
        "module_args": {
            "_raw_params": "/root/racadm.sh root 3KWAAMK9W3PN 9.42.98.136 coreos.iso",
            "_uses_shell": true,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "msg": "non-zero return code",
    "rc": 126,
    "start": "2021-01-20 03:57:39.710551",
    "stderr": "/bin/sh: /root/racadm.sh: cannot execute binary file: Exec format error",
    "stderr_lines": [
        "/bin/sh: /root/racadm.sh: cannot execute binary file: Exec format error"
    ],
    "stdout": "",
    "stdout_lines": []
}
```

### Issue 12. 
How to add more worker nodes not specified. (This is addressed in the latest scripts)
```
# bmc_type based on your HW it should be one of the following values [Dell, SuperMicro].
# You can add more bmc_type however you will need to add the bmc implementation at add-remote-worker role.
bmc_type=Dell
bmc_address=9.42.98.133
bmc_user="root"
bmc_password="F9EAHWEV439A"
```

### Issue 13. 
I found these IP address defined in the inventory hosts, was not used/assigned. Instead they got the IP from dhcp server. (This feature is anyhow removed from the latest scripts)
```
# You can ignore this part if you have dhcp configured in your env.
[master_vms]
9.42.98.153
9.42.98.154
9.42.98.155
```

**Comment by yrobla** [comment]
Can you divide that into independent issues, and confirm if some of them have already been resolved? Thanks

**Comment by maheshd2** [comment]
I have opened new issues for the items which needs attention. And hence this issue can be marked closed.
